src/batch/src/executor2/hop_window.rs

```diff
@@ -0,0 +1,305 @@
+            .exact_div(&window_slide)
+            .and_then(|x| NonZeroUsize::new(usize::try_from(x).ok()?))
+            .ok_or_else(|| {
+                RwError::from(ErrorCode::InternalError(format!(
+                    "window_size {} cannot be divided by window_slide {}",
+                    window_size, window_slide
+                )))
+            })?
+            .get();
+
+        let schema = self.schema;
+        let time_col_data_type = schema.fields()[time_col_idx].data_type();
+        let time_col_ref = InputRefExpression::new(time_col_data_type, self.time_col_idx).boxed();
+
+        let window_slide_expr =
+            LiteralExpression::new(DataType::Interval, Some(ScalarImpl::Interval(window_slide)))
+                .boxed();
+
+        // The first window_start of hop window should be:
+        // tumble_start(`time_col` - (`window_size` - `window_slide`), `window_slide`).
+        // Let's pre calculate (`window_size` - `window_slide`).
+        let window_size_sub_slide = window_size.checked_sub(&window_slide).ok_or_else(|| {
+            RwError::from(ErrorCode::InternalError(format!(
+                "window_size {} cannot be subtracted by window_slide {}",
+                window_size, window_slide
+            )))
+        })?;
+        let window_size_sub_slide_expr = LiteralExpression::new(
+            DataType::Interval,
+            Some(ScalarImpl::Interval(window_size_sub_slide)),
+        )
+        .boxed();
+        let hop_expr = new_binary_expr(
+            expr_node::Type::TumbleStart,
+            risingwave_common::types::DataType::Timestamp,
+            new_binary_expr(
+                expr_node::Type::Subtract,
+                DataType::Timestamp,
+                time_col_ref,
+                window_size_sub_slide_expr,
+            ),
+            window_slide_expr,
+        );
+
+        #[for_await]
+        for data_chunk in child.execute() {
+            let data_chunk = data_chunk?;
+            let hop_start = hop_expr.eval(&data_chunk)?;
+            let hop_start_chunk = DataChunk::new(vec![Column::new(hop_start)], None);
+            let (origin_cols, visibility) = data_chunk.into_parts();
+            // SAFETY: Already compacted.
+            assert!(visibility.is_none());
+            for i in 0..units {
+                let window_start_offset = window_slide.checked_mul_int(i).ok_or_else(|| {
+                    RwError::from(ErrorCode::InternalError(format!(
+                        "window_slide {} cannot be multiplied by {}",
+                        window_slide, i
+                    )))
+                })?;
+                let window_start_offset_expr = LiteralExpression::new(
+                    DataType::Interval,
+                    Some(ScalarImpl::Interval(window_start_offset)),
+                )
+                .boxed();
+                let window_end_offset =
+                    window_slide.checked_mul_int(i + units).ok_or_else(|| {
```

Comment:
We create expression for each input chunk?

Reply:
Emmm. Maybe I can create an array with size of `units`.

Reply:
done.